spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/pubsub/GcpPubSubAutoConfiguration.java

```diff
@@ -167,10 +167,12 @@ public Executor pubSubAcknowledgementExecutor() {
 	@ConditionalOnMissingBean
 	public PubSubSubscriberTemplate pubSubSubscriberTemplate(SubscriberFactory subscriberFactory,
 			ObjectProvider<PubSubMessageConverter> pubSubMessageConverter,
-			@Qualifier("pubSubAcknowledgementExecutor") Executor executor) {
+			@Qualifier("pubSubAsynchronousPullExecutor") ObjectProvider<Executor> asyncPullExecutor,
+			@Qualifier("pubSubAcknowledgementExecutor") Executor ackExecutor) {
```

Comment:
If we're using the `ObjectProvider` for the `asyncPullExecutor`, why doesn't it apply to `ackExecutor`?
Both approaches might be valid, but we should try to be consistent.

Reply:
I did not change that into a `ObjectProvider` because there is always a `pubSubAcknowledgementExecutor` bean in the spring context (see the `pubSubAcknowledgementExecutor()` bean definition method above this method). So, having a  `ObjectProvider` is pointless as it will always have a bean. Agree?

Side note: you can wonder why we need the `pubSubAcknowledgementExecutor` bean at all, and why the default ackExecutor, defined in `PubSubSubscriberTemplate`, doesn't suffice. Currently it is a `Executors.newSingleThreadExecutor()`, which doesn't scale, I guess, and that is why the `pubSubAcknowledgementExecutor` bean defines a ThreadPoolTaskExecutor that can have more threads, depending on the `...subscriber.max-acknowledgement-threads` property. 

We could also just change the default `Executors.newSingleThreadExecutor()` in the `PubSubSubscriberTemplate` in to  a `Runnable::run` and the `pubSubAcknowledgementExecutor` bean could only be created if a specific `...subscriber.max-acknowledgement-threads` is defined in the properties (that way we stay somewhat backwards compatible). That way it is  in line how we do callback processing for both ack responses and message pull responses. I haven't done any such changes though, as that would be a big change in how the PubSubSubscriberTemplate works by default for ack responses. Also it is unrelated to my ticket.

Reply:
I think you're right, this raises more questions about the introduction of the `pubSubAcknowledgementExecutor` bean rather than the use of the `ObjectProvider` for `asyncPullExecutor`. The change was made in #1349 over a year ago and the `...subscriber.max-acknowledgement-threads` was never really documented in the refdocs. So, it might be OK to just remove it, and replace with the `Runnable::run` default in the template itself. However, it looks like there were some performance issues with the direct executor. In any case, like you said, it's irrelevant to your ticket. @elefeint Should we file a separate ticket to investigate the `pubSubAcknowledgementExecutor`?

Reply:
Yes, let's create a different issue, and I agree that using `Runnable::run` by default is the right thing to do here.

Reply:
#2245